htlcswitch/hop/payload.go

```diff
@@ -256,33 +256,48 @@ func ValidateParsedPayloadTypes(parsedTypes tlv.TypeMap,
 	_, hasNextHop := parsedTypes[record.NextHopOnionType]
 	_, hasMPP := parsedTypes[record.MPPOnionType]
 	_, hasAMP := parsedTypes[record.AMPOnionType]
+	_, hasEncryptedData := parsedTypes[record.EncryptedDataOnionType]
 
-	switch {
+	// All cleartext hops (including final hop) and the final hop in a
+	// blinded path require the forwading amount and expiry TLVs to be set.
+	needFwdInfo := isFinalHop || !hasEncryptedData
+
+	// No blinded hops should have a next hop specified, and only the final
+	// hop in a cleartext route should exclude it.
+	needNextHop := !(hasEncryptedData || isFinalHop)
 
+	switch {
 	// All hops must include an amount to forward.
-	case !hasAmt:
+	case needFwdInfo && !hasAmt:
```

Comment:
There's a corner case here.  `hasAmt` will be true if the TLV is present but 0.

The final hop fuzz test still fails for this input:  `[]byte("\x02\x00\x04\x00")`.  Decoding succeeds because the TLV is present but 0.  Then encoding fails because the amount is 0.

I'm not sure if we should fail to decode 0 amount or locktime here or if we should check that at a higher level somewhere (maybe we already do that?).  Failing during decoding would make the fuzz test start passing.  Otherwise, if we fail at a higher level, we'll need to except this corner case in the fuzz test.

Reply:
> I'm not sure if we should fail to decode 0 amount or locktime here or if we should check that at a higher level somewhere (maybe we already do that?).

Yeah, we check that the HTLC's amount/expiry are sane later on. 

> Otherwise, if we fail at a higher level, we'll need to except this corner case in the fuzz test.

Somewhat torn here. I don't think that failing during decoding for a TLV that is present but zero is the right approach (value semantics don't belong in decoding),  but it seems like a shame to exclude all of these cases from fuzzing - iiuc we'd need to not fail on `ErrMissingField` because `amt`/`expiry` or even `chanid` could be present but zero?